.github/workflows/benchmark.yml

```diff
@@ -0,0 +1,33 @@
+name: Benchmark
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+env:
+  DEFAULT_GO_VERSION: 1.17
+jobs:
+  benchmark:
+    name: Benchmarks
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-go@v2
+        with:
+          go-version: ${{ env.DEFAULT_GO_VERSION }}
+      - name: Run benchmarks
+        run: make test-bench | tee output.txt
+      - name: Download previous benchmark data
+        uses: actions/cache@v2
+        with:
+          path: ./benchmarks
+          key: ${{ runner.os }}-benchmark
```

Comment:
AFAIK we need to scope this by branch or only have the `main` branch update the cache somehow. Otherwise all PRs will update this cache and the "baseline" will fluctuate with each PR.

Reply:
If we do so, PRs won't be able to fetch the cache from the main branch and will always start from fresh.
Also, based on [this doc](https://docs.github.com/en/actions/advanced-guides/caching-dependencies-to-speed-up-workflows#restrictions-for-accessing-a-cache), it seems PRs read their cache from `main`, but write write it to a location which is never used by `main`.

Reply:
However, it also means a PR could self-block itself if it runs the benchmark very quickly once, and then proceeds to running it more slowly in further pushes.
Ideally, we should be able to skip storing the cache for non-pushes. But there's no solution to that at the moment. So I've disabled running those for prs for now :(